### PR TITLE
agent: making agent internals private

### PR DESCRIPTION
**Pull request by Julio-Guerra**
Move the agent internals into a package named `internal`, which makes it importable from parent packages only, and thus impossible to import from outside packages.

It is therefore required to separate the SDK dependency to the agent internals by defining a Go interface including everything it expects from the agent. The agent automatically "sets itself" on initialization. When the agent is not included, the SDK uses a disabled agent, which no-ops everything.

A consequence is that it is easy now to unit-test the SDK with 100% coverage by simply mocking the agent and checking everything.

Related to SQR-5192, SQR-5204, SQR-5242 and SQR-5262.

### CODE REVIEW

**Comment by vdeturckheim** [comment]
@Julio-Guerra does that worksd on all version of Go?

File: agent/types/types.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	"net/http"
+	"time"
+)
+
+type Agent interface {
+	NewRequestRecord(req *http.Request) RequestRecord
+	GracefulStop()
+}
+
+type RequestRecord interface {
```
Comment (vdeturckheim) [inline agent/types/types.go]: I would not expose the concept of Request Record and rather expose that as SDK

File: sdk/agent.go [hunk 2]
```diff
@@ -1,7 +1,61 @@
...
 	agent.GracefulStop()
 }
+
+type disabledAgent struct {
+}
+
+func (_ disabledAgent) GracefulStop() {
+}
+
+func (a disabledAgent) NewRequestRecord(_ *http.Request) types.RequestRecord {
```
Comment (vdeturckheim) [inline sdk/agent.go]: This is user facing right?

**Comment by vdeturckheim** [comment]
I'm under the impression we could do more in coverage right?

File: agent/types/types.go (hunk 1 above)
Comment (Julio-Guerra) [reply agent/types/types.go]: I need a name for the structure we attach to HTTP contextes. NewRelic called their structure `Transaction`, and for us I chose "request record of of events", ie. `HTTPRequestRecord`.

There is only one type exposed to users and it is `HTTPRequestRecord`. The others are exported by the agent but meant to be used by the SDK, and indeed there are no way to use them is the SDK interface.

File: sdk/agent.go (hunk 2 above)
Comment (Julio-Guerra) [reply sdk/agent.go]: No, SDK only.

**Comment by Julio-Guerra** [comment]
> @Julio-Guerra does that worksd on all version of Go?

:+1:  Planned and tracked this in https://sqreen.atlassian.net/browse/SQR-5703

**Comment by Julio-Guerra** [comment]
> I'm under the impression we could do more in coverage right?

Yes, and I am going to address that in an upcoming PRs. This one makes 100% coverage of the `sdk` layer (updated after a push force - and codecov doesn't update its comments - and jenkins is dead right now so no new coverage update).

**Pull request merged**